misk-metrics-digester/src/main/kotlin/misk/metrics/digester/Registry.kt

```diff
@@ -0,0 +1,109 @@
+package misk.metrics.digester
+
+import misk.metrics.Histogram
+import misk.metrics.HistogramRecordMetric
+import misk.metrics.HistogramRegistry
+import java.util.Objects
+import java.util.concurrent.ConcurrentHashMap
+
+/** DigestMetric contains the contents of an individual metrics within a histogram */
+class DigestMetric(
+  val digest: SlidingWindowDigest<*>,
+  val labels: List<String>
+) {
+  /** Adds an observation to the metric */
+  fun observe(value: Double) {
+    digest.observe(value)
+  }
+}
+
+class TDigestHistogramRecordMetric constructor(private val metric: DigestMetric) :
+    HistogramRecordMetric {
+  override fun observe(duration: Double) {
+    metric.observe(duration)
+  }
+}
+
+/** TDigestHistogram stores histograms which records metrics */
+class TDigestHistogram<T : TDigest<T>> constructor(
+  val name: String,
+  val help: String,
+  val quantiles: List<Double>,
+  private val labelNames: List<String>,
+  private val tDigest: () -> SlidingWindowDigest<T>
+) : Histogram {
+
+  private val metrics = ConcurrentHashMap<Int, DigestMetric>()
+
+  /** Add all labels passed to the histogram */
+  init {
```

Comment:
Remove this. It is unintentionally creating metrics.

Say label names given are `[a, b]`. This creates two metrics with label values `{a->a}` and `{b->b}`

Reply:
Do we not want this at all? This is to maintain a similar approach to Prometheus histogram metrics.
```  init {
      labels(*labelNames as Array<out String>)
  } ```

Reply:
No, because we don't want to mix up label names with label values.

```
// This does not create any metrics _yet_.
// However, it means that every metric derived from it will have exactly two labels
// and the _names_ of those labels will always be label1 and label2.
val hist = registry.newHistogram("yourmetric", "help", listOf("label1", "label2"), mapOf(...))

// This means a metric named "yourmetric" with labels {"label1":"value1", "label2":"value2"} is sent to Prometheus
hist.labels("value1", "value2").observe(1.23)

// This means a metric named "yourmetric" with labels {"label1":"x", "label2":"y"} is sent to Prometheus
hist.labels("x", "y").observe(1.23)
```

Your init here is effectively doing:
```
// Invalid because no observed metric has a _value_ of label1 and also because a value for label2 is omitted.
hist.labels("label1")

// Invalid because no observed metric has a _value_ of label2 and also because a value for label1 is omitted.
hist.labels("label2")
```